train/uft/tuning.py

```diff
@@ -0,0 +1,130 @@
+from torch.utils.data import DataLoader
+from pathlib import Path
+import sys
+import os
+import yaml
+import optuna
+from fire import Fire
+
+from train import train_loop
+
+sys.path.append(str(Path(__file__).parent.parent.parent))
+from evaluation.evaluate import evaluate
+from utils.preprocess import (
+    create_train_and_validation_unpaired_datasets,
+    create_evaluation_dataset
+)
+from utils.eval_plots import (
+    plot_and_save_metrics,
+    plot_and_save_metric_distribution,
+    calc_and_save_average_metrics,
+    save_aggregated_data,
+)
+from utils.utils import (
+    create_unique_save_dir,
+    update_key_if_new_value_is_not_none,
+    create_dir_with_hyperparam_name,
+)
+from utils.train_plots import (
+    save_denoiser_and_discriminator_loss_records_in_csv,
+    plot_train_and_val_loss_curves,
+)
+
+
+def load_directory_from_yml(config_path: str | Path) -> dict:
+
+    with open(config_path) as ymlfile:
+        config = yaml.safe_load(ymlfile)
+
+    return {
+        "checkpoint_path": config.get("checkpoint_path"),
+        "images_dir": config.get("images_dir"),
+        "save_dir": config.get("save_dir"),
+        "eval_dir": config.get("eval_dir")
+    }
+
+def get_params(trial):
+    # Generate the optimizers.
+    params = {
+            'denoiser_adam_lr': trial.suggest_float('denoiser_adam_lr', 1e-6, 1e-2, log = True), 
+            'discriminator_adam_lr': trial.suggest_float('discriminator_adam_lr', 1e-6, 1e-2, log = True), 
+            'denoiser_loss_b1' : trial.suggest_int('denoiser_loss_b1', 0.1, 0.9),
+            'denoiser_loss_b2' : trial.suggest_float('denoiser_loss_b2',0.1, 0.9),
+            'print_loss_interval': trial.suggest_int('print_loss_interval', 50, 500),
+            'denoiser_adversarial_loss_clip_min': trial.suggest_float('denoiser_adversarial_loss_clip_min', 0.0, 1.0),
+            'denoiser_adversarial_loss_clip_max': trial.suggest_float('denoiser_adversarial_loss_clip_max', 0.0, 1.0),
+            'calc_eval_loss_interval': trial.suggest_int("calc_eval_loss_interval", 50, 500),
+    }
+
+    return params
+
+def objective (trial):
+    config_path = Path(__file__).parent / "config.yml"
+    images_dir =  load_directory_from_yml(config_path)["images_dir"]
```

Comment:
i think you are calling load_directory_from_yml multiple times in this function. can you call once and store all the variables at the same time? (you are accessing the disk everytime to load config)

Reply:
ok can